File: algo/src/main/java/com/hartwig/actin/algo/calendar/HistoricDateProvider.java
package com.hartwig.actin.algo.calendar;

import java.time.LocalDate;

import com.hartwig.actin.clinical.datamodel.ClinicalRecord;

import org.jetbrains.annotations.NotNull;

class HistoricDateProvider implements ReferenceDateProvider {

    @NotNull
    private final LocalDate historicDate;

    @NotNull
    public static HistoricDateProvider fromClinical(@NotNull ClinicalRecord clinical) {
        return new HistoricDateProvider(clinical.patient().registrationDate().plusWeeks(3));
    }

    private HistoricDateProvider(@NotNull final LocalDate historicDate) {
        this.historicDate = historicDate;
    }

    @NotNull
    @Override
    public LocalDate date() {
        return null;
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/calendar/HistoricDateProvider.kt
package com.hartwig.actin.algo.calendar

import com.hartwig.actin.datamodel.clinical.PatientDetails
import java.time.LocalDate

internal class HistoricDateProvider private constructor(private val historicDate: LocalDate) : ReferenceDateProvider {

    override fun date(): LocalDate {
        return historicDate
    }

    override val isLive: Boolean
        get() = false

    companion object {
        fun fromPatientDetails(patientDetails: PatientDetails): HistoricDateProvider {
            val historicDate = patientDetails.registrationDate.plusWeeks(3)
            val currentDate = LocalDate.now()
            val effectiveDate = if (currentDate.isBefore(historicDate)) currentDate else historicDate
            return HistoricDateProvider(effectiveDate)
        }
    }
}